query/src/main/java/org/infinispan/query/impl/PartitionHandlingSupport.java

```diff
@@ -0,0 +1,39 @@
+package org.infinispan.query.impl;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.configuration.cache.ClusteringConfiguration;
+import org.infinispan.partitionhandling.AvailabilityMode;
+import org.infinispan.partitionhandling.PartitionHandling;
+import org.infinispan.query.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * @since 9.3
+ */
+public final class PartitionHandlingSupport {
+
+   private static final Log LOGGER = LogFactory.getLog(CacheQueryImpl.class, Log.class);
+
+   private final boolean isClustered;
+   private final PartitionHandling partitionHandling;
+   private final AdvancedCache<?, ?> cache;
+
+   public PartitionHandlingSupport(AdvancedCache<?, ?> cache) {
+      this.cache = cache;
+      ClusteringConfiguration clusteringConfiguration = cache.getCacheConfiguration().clustering();
+      this.isClustered = clusteringConfiguration.cacheMode().isClustered();
+      this.partitionHandling = isClustered ? clusteringConfiguration.partitionHandling().whenSplit() : null;
+   }
+
+   public void checkCacheAvailable() {
+      if (!isClustered) return;
+
+      AvailabilityMode availability = cache.getAvailability();
+      if (availability == AvailabilityMode.AVAILABLE) return;
+
+      if (partitionHandling != PartitionHandling.ALLOW_READ_WRITES) {
+         throw LOGGER.partitionDegraded();
```

Comment:
I tend to agree that we shouldn't allow queries in degraded mode when configured with `ALLOW_READS`, in fact I wonder if we shouldn't go further and warn the user on startup that they shouldn't use `ALLOW_READS` with indexed query, because it behaves like `DENY_READS_WRITES` and not like they might expect from the `ALLOW_READS` javadoc/user guide/xsd. Or maybe we should ignore the fact that query is a separate module and document how `ALLOW_READS` behaves for query everywhere?